Add MLMC example [feature/mlmc-manager]

barker29 opened the pull request:
Previously, we really only had multilevel sampling without any real multilevel Monte Carlo. The previously named `mlmc.cpp` is now `samplesolve.cpp`, and we do some very basic multilevel Monte Carlo in `qoi.cpp`.

barker29 committed:
rename example mlmc.cpp -> samplesolve.cpp

barker29 committed:
MLMCManager: add object and QoI

barker29 committed:
finitevolume: use formal QuantityOfInterest class in this example

This is overkill here, but starts to give an idea of how the
class might be useful.

barker29 committed:
WIP: qoi: draft the example, it compiles but is untested

barker29 committed:
Slightly prettier pictures in qoi.cpp

barker29 committed:
qoi: slightly better defaults

barker29 committed:
qoi: better documentation and a new test

barker29 committed:
QuantityOfInterest: change interfaces to be more general and flexible

barker29 committed:
qoi: add a simple script to visualize 2D output from qoi.cpp

chakshinglee commented on src/MLMCManager.cpp:
```diff
@@ -0,0 +1,450 @@
+    {
+        out += v;
+    }
+    return out;
```
this function may be simplified by calling `std::accumulate`:
```
return std::accumulate(mean_.begin(), mean_.end(), 0.0);
```

chakshinglee commented on src/MLMCManager.cpp:
```diff
@@ -0,0 +1,450 @@
+    return out;
+}
+
+void MLMCManager::CoarseSample(bool verbose)
```
This shares a lot of code with `MLMCManager::FineSample`. 

Maybe we can think of `MLMCManager` and `Upscale` as two different objects in that `Upscale` provides a "M-level" hierarchy, `MLMCManager` can do a "m-level" MLMC simulation for any "m" <= "M". This way, at the coarsest level "m-1" we will always just call this function (in the case of fine level MC, `num_levels_` = 0, so this function will pretty much do what `FineSample` does, except the rhs, which is commented below).

For any level "k" < "m-1", the CorrectionSample will be call.

chakshinglee commented on src/MLMCManager.cpp:
```diff
@@ -0,0 +1,450 @@
+    sampler_.NewSample();
+    auto coarse_coefficient = sampler_.GetCoefficient(level);
+    fvupscale_.RescaleCoefficient(level, coarse_coefficient);
+    mfem::BlockVector rhs_coarse = RestrictToLevel(fvupscale_, level, rhs_fine_);
```
Whenever a new coefficient is sampled, `rhs_fine_` and `rhs_coarse` are not changed. So this restriction step can be performed once, and the rhs for all levels are stored for repeated use (both here and `CorrectionSample`.

chakshinglee commented on src/MLMCManager.hpp:
```diff
@@ -0,0 +1,247 @@
+       It might make sense to combine coefficient, pressure in some
+       kind of "state" object (that may even manage the solve), for now
+       we will just require both all the time even if you use just one.
+    */
```
Seems `coefficient` is not used in any derived class of `QuantityOfInterest`, are you thinking there is potentially some other quantity of interest that needs coefficient to evaluate?

Or you are thinking `QuantityOfInterest` as a map that takes a sample of `coefficient` and return the value of a certain quantity of interest (like you said, manage the solve as well)? In that case we probably dont need pressure solution because that depends on `coefficient`.

barker29 replied:
I have a use case in mind for including coefficient, but you are right it is not implemented now. I can take it out but I think the abstract interface makes sense to include it.

chakshinglee replied:
I see, let this stays as it is right now then.

chakshinglee commented on src/Upscale.cpp:
```diff
@@ -183,7 +183,7 @@ void Upscale::SolveAtLevel(int level, const mfem::Vector& x, mfem::Vector& y) co
 mfem::Vector Upscale::SolveAtLevel(int level, const mfem::Vector& x) const
 {
-    mfem::Vector coarse_vect = GetVector(1);
+    mfem::Vector coarse_vect = GetVector(level);
```
Good catch! How did this function not break before? Maybe this is never called for `level !=1`.

chakshinglee commented on src/MLMCManager.cpp:
```diff
@@ -0,0 +1,450 @@
+    varsum_[fine_level] += scale * ((double) sample_count_[fine_level]) *
+                           ((fineq - upscaledq) - mean_[fine_level]) * ((fineq - upscaledq) - mean_[fine_level]);
+    mean_[fine_level] += scale * ((fineq - upscaledq) - mean_[fine_level]);
+    cost_[fine_level] += scale * (temp_cost - cost_[fine_level]);
```
this few lines (318-322) can be grouped as a function and be called in both `CoarseSample` and and `CorrectionSample`, with input parameters `level` and qoi (for here it is `fineq - upscaledq`, for `CoarseSample` it is QoI of that level).

chakshinglee commented on src/MLMCManager.cpp:
```diff
@@ -0,0 +1,450 @@
+        // does it make sense to log some information, like the QoI, to go with the visualization?
+        // (conceivably you could even ask qoi_ to make a special picture)
+
+        // for more informative visualization (is it more sensible to do this here or in the python?)
```
If you only want log scale for visualization in GLVis, there is an option in GLVis (key "L" if I remember correctly)

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,251 @@
+
+    std::unique_ptr<MultilevelSampler> sampler;
+    const int seed = argseed + myid;
+    sampler = make_unique<PDESampler>(
```
Minor: use of pointer is not needed here.

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,251 @@
+            mlmc.CorrectionSample(0, verbose);
+        }
+    }
+    else
```
I think this case covers `upscale_param.max_levels == 2` as well

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,251 @@
+            std::cout << "---\nChoose sample " << sample << "\n---" << std::endl;
+
+        mlmc.BestSample(verbose);
+    }
```
It probably makes sense to wrap L189-243 into a function 
(say, `MLMCManager::Simulate(int num_levels, bool verbose)`, where `num_levels` is the number of levels of the MLMC simulation, which can be different from the number of levels of `Upscale` object that MLMCManager is constructed from)

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,251 @@
+                   "Spatial dimension of simulation.");
+    double cell_volume = 10000.0;
+    args.AddOption(&cell_volume, "--volume", "--volume",
+                   "Volume of typical cell (for PDE sampler).");
```
Should this be obtained from `DarcyProblem::CellVolume` or `pmesh_`? It should match with `scale`.

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,251 @@
+    mfem::Array<int> ess_attr(dimension == 3 ? 6 : 4);
+    DarcyProblem fvproblem(*pmesh, ess_attr);
+    Graph graph = fvproblem.GetFVGraph();
+    Upscale upscale(graph, upscale_param);
```
if `ess_attr` is not provided, no elimination of edofs will be performed, and `Upscale` will treat it as a regular graph (no weird edges). I am not sure what is the final discrete problem that it represents. Similarly for `PDESampler`. 

Also, `HybridSolver` breaks because of the mismatch in graph boundary and `ess_attr_` (which is `nullptr` if not provided. For these two reasons I suggest to put `ess_attr` as one of the parameters (in that case you probably want to give an explicit value for `ess_attr`.

chakshinglee reviewed (commented):
This new feature looks so cool! 

Other than the individual comments, I am thinking maybe it makes sense to give some parameters to the `MLMCManager` to make it something analogous to linear solvers, one is maximum number of samples (it is something like `fine_samples` etc in `qoi.cpp`), the other one is tolerance for "approximated" mean square error (which ignores discretization error), and the simulation will stop when the error reaches tolerance. If tolerance is zero, then the simulation is only controlled by number of samples.

barker29 committed:
MLMCManager: add FixedLevelSample() which is called by both FineSample() and CoarseSample()

barker29 committed:
MLMCManager: remove FineSample, CoarseSample

barker29 committed:
MLMCManager: add UpdateStatistics() utility routine to reduce repeated code

barker29 committed:
MLMCManager: this object now can have a different number of levels from its Upscale parameter

barker29 committed:
qoi: Simplify sampling loop

barker29 committed:
MLMCManager: store rhs_ at all levels in constructor instead of interpolating all the time

barker29 committed:
qoi: get cell_volume from DarcyProblem instead of command line

barker29 committed:
MLMCManager: add Simulate() method to run the whole algorithm

barker29 committed:
test: rebaseline for loop order

barker29 committed:
qoi: Fix boundary conditions and rebaseline tests.

This change actually makes it solve quite a different problem, so
the changes to the tests are significant, but I am still confident
this example is solving the problem and getting a good Monte Carlo
estimate.

barker29 commented:
Thanks for all your comments, they were very helpful!

I think I've addressed all of them - please take another quick look and approve/decline.

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,205 @@
+    {
+        functional(i) = dscale;
+    }
+    mfem::Vector dummy;
```
this line can be removed

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,205 @@
+    mfem::Vector functional(rhs_fine.GetBlock(1));
+    functional = 0.0;
+    double dscale = 1.0 / problem_size;
+    for (int i = 0; i < problem_size; ++i)
```
My understanding is that your are defining a functional to compute some sort of average pressure on the bottom boundary (as stated in the description). This definition is not computing that in 3D? But we can measure anything, so it doesn't really matter, it's up to you whether to change this or not.

chakshinglee reviewed (approved):
@barker29 I just had some minor comments. The rest look great to me, thanks!

chakshinglee commented on examples/qoi.cpp:
```diff
@@ -0,0 +1,205 @@
+
+    mfem::BlockVector rhs_fine(upscale.GetBlockVector(0));
+    rhs_fine.GetBlock(0) = 0.0;
+    rhs_fine.GetBlock(1) = 0.0;
```
these two lines can be combined to `rhs_fine = 0.0;`

barker29 committed:
qoi: minor improvements
